### PR TITLE
docs: replace <pre> with ```

### PR DESCRIPTION
**Pull request by gpanders**
Still a few things not quite right, but this is probably ~90% correct already.

### CODE REVIEW

**Comment by gpanders** [comment]
@justinmk I don't think we can completely eliminate `<pre>` yet. In cases where we have a code sample inside of a list, a ```` ``` ```` code block doesn't work (at least, I can't figure out how to make it work). Example: https://github.com/neovim/neovim/blob/27a566f3f8e07a4cebb426674800bdf9a7f4f222/runtime/lua/vim/diagnostic.lua#L600-L610

We also use `<pre>help` for some docs that are literal help text. I didn't touch any of that.

File: runtime/doc/api.txt [hunk 1]
```diff
@@ -3054,14 +3066,25 @@ nvim_open_win({buffer}, {enter}, {*config})                  *nvim_open_win()*
     could let floats hover outside of the main window like a tooltip, but this
     should not be used to specify arbitrary WM screen positions.
 
-    Example (Lua): window-relative float >lua
-        vim.api.nvim_open_win(0, false,
-          {relative='win', row=3, col=3, width=12, height=3})
+    Example (Lua): window-relative float>lua
```
Comment (gpanders) [inline runtime/doc/api.txt:3069]: Missing space before `>lua`

File: runtime/doc/api.txt [hunk 2]
```diff
@@ -3325,24 +3349,27 @@ nvim_create_autocmd({event}, {*opts})                  *nvim_create_autocmd()*
     Creates an |autocommand| event handler, defined by `callback` (Lua function or Vimscript function name string) or `command` (Ex command string).
 
     Example using Lua callback: >lua
+
         vim.api.nvim_create_autocmd({"BufEnter", "BufWinEnter"}, {
           pattern = {"*.c", "*.h"},
           callback = function(ev)
-            print(string.format('event fired: %s', vim.inspect(ev)))
+            print(string.format('event fired: \%s', vim.inspect(ev)))
```
Comment (gpanders) [inline runtime/doc/api.txt:3356]: `%` being escaped for some reason

File: runtime/doc/lsp.txt [hunk 3]
```diff
@@ -969,10 +969,8 @@ start_client({config})                                *vim.lsp.start_client()*
                     the `cmd` process. Not related to `root_dir`.
                   • cmd_env: (table) Environment flags to pass to the LSP on
                     spawn. Must be specified using a table. Non-string values
-                    are coerced to string. Example: >
-
-                       { PORT = 8080; HOST = "0.0.0.0"; }
-<
+                    are coerced to string. Example: ` { PORT = 8080; HOST =
+                    "0.0.0.0"; } `
```
Comment (gpanders) [inline runtime/doc/lsp.txt:973]: Need to use `<pre>` here

File: runtime/doc/lsp.txt [hunk 4]
```diff
@@ -1242,13 +1242,9 @@ format({options})                                       *vim.lsp.buf.format()*
...
-                     Clients matching the predicate are included. Example:               • >lua
-
-               -- Never request typescript-language-server for formatting
-               vim.lsp.buf.format {
-                 filter = function(client) return client.name ~= "tsserver" end
-               }
-<
+                     Clients matching the predicate are included. Example:               • `lua -- Never request typescript-language-server for
+                     formatting vim.lsp.buf.format { filter = function(client)
+                     return client.name ~= "tsserver" end } `
```
Comment (gpanders) [inline runtime/doc/lsp.txt:1247]: Need to use `<pre>` here

File: runtime/doc/lsp.txt [hunk 5]
```diff
@@ -1456,8 +1452,11 @@ refresh()                                         *vim.lsp.codelens.refresh()*
...
     It is recommended to trigger this using an autocmd or via keymap.
 
-    Example: >vim
-      autocmd BufEnter,CursorHold,InsertLeave <buffer> lua vim.lsp.codelens.refresh()
+    Example:
+
+    >vim
+
+        autocmd BufEnter,CursorHold,InsertLeave <buffer> lua vim.lsp.codelens.refresh()
 <
```
Comment (gpanders) [inline runtime/doc/lsp.txt:1460]: `>vim` marker got bumped to a newline

File: runtime/doc/lsp.txt [hunk 6]
```diff
@@ -1586,14 +1585,16 @@ hover({_}, {result}, {ctx}, {config})               *vim.lsp.handlers.hover()*
...
+    parameter is highlighted with |hl-LspSignatureActiveParameter|.
+
+    >lua
+
+        vim.lsp.handlers["textDocument/signatureHelp"] = vim.lsp.with(
+          vim.lsp.handlers.signature_help, {
+            -- Use a sharp border with `FloatBorder` highlights
+            border = "single"
+          }
+        )
```
Comment (gpanders) [inline runtime/doc/lsp.txt:1597]: `>lua` bumped to a newline

File: runtime/doc/lua.txt [hunk 7]
```diff
@@ -582,17 +582,25 @@ VIM.HIGHLIGHT                                                  *vim.highlight*
...
+
     au TextYankPost * silent! lua vim.highlight.on_yank {higroup="IncSearch", timeout=150}
+
 <
 
-If you want to exclude visual selections from highlighting on yank, use: >vim
+If you want to exclude visual selections from highlighting on yank, use:
+
+>vim
+
```
Comment (gpanders) [inline runtime/doc/lua.txt:601]: Newline

File: runtime/doc/lua.txt [hunk 8]
```diff
@@ -689,18 +697,18 @@ vim.diff({a}, {b}, {opts})                                        *vim.diff()*
...
-         -- @ -1 +1,2 @
-         -- -a
-         -- +b
-         -- +c
+        vim.diff('a\\n', 'b\\nc\\n')
+        -- =>
+        -- @@ -1 +1,2 @@
+        -- -a
+        -- +b
+        -- +c
```
Comment (gpanders) [inline runtime/doc/lua.txt:705]: Additional backslashes.

File: runtime/doc/lua.txt [hunk 9]
```diff
@@ -782,10 +790,11 @@ vim.json.decode({str}, {opts})                             *vim.json.decode()*
...
     Example: >lua
 
-     :lua vim.print(vim.json.decode('{"bar":[],"foo":{},"zub":null}'))
-     --> { bar = {}, foo = vim.empty_dict(), zub = vim.NIL }
+        vim.print(vim.json.decode('{"bar":[],"foo":{},"zub":null}'))
+        -- { bar = {}, foo = vim.empty_dict(), zub = vim.NIL }
+<
 
-    < Parameters: ~ • {str} Stringified JSON data. • {opts} Options map keys: •
+    Parameters: ~ • {str} Stringified JSON data. • {opts} Options map keys: •
```
Comment (gpanders) [inline runtime/doc/lua.txt:797]: This didn't happen in this PR, but this isn't right

File: runtime/doc/lua.txt [hunk 10]
```diff
@@ -1423,7 +1440,10 @@ vim.bo                                                                *vim.bo*
 
     Note: this is equivalent to both `:set` and `:setlocal`.
 
-    Example: >lua
+    Example:
+
+    >lua
+
```
Comment (gpanders) [inline runtime/doc/lua.txt:1446]: newline

File: runtime/doc/lua.txt [hunk 11]
```diff
@@ -1451,7 +1474,10 @@ vim.go                                                                *vim.go*
     option value and thus is mostly useful for use with |global-local|
     options.
 
-    Example: >lua
+    Example:
+
+    >lua
+
```
Comment (gpanders) [inline runtime/doc/lua.txt:1480]: newline

File: runtime/doc/lua.txt [hunk 12]
```diff
@@ -1463,7 +1489,8 @@ vim.o                                                                  *vim.o*
     Note: this works on both buffer-scoped and window-scoped options using the
     current buffer and window.
 
-    Example: >lua
+    Example:>lua
```
Comment (gpanders) [inline runtime/doc/lua.txt:1492]: missing space before `>lua`

File: runtime/doc/lua.txt [hunk 13]
```diff
@@ -1478,7 +1505,10 @@ vim.wo                                                                *vim.wo*
     Note: only {bufnr} with value `0` (the current buffer in the window) is
     supported.
 
-    Example: >lua
+    Example:
+
+    >lua
+
```
Comment (gpanders) [inline runtime/doc/lua.txt:1511]: newline

File: runtime/doc/lua.txt [hunk 14]
```diff
@@ -2527,6 +2547,13 @@ vim.ui.input({opts}, {on_confirm})                            *vim.ui.input()*
                       typed (it might be an empty string if nothing was
                       entered), or `nil` if the user aborted the dialog.
 
+    >lua
+
+        vim.ui.input({ prompt = 'Enter value for shiftwidth: ' }, function(input)
+            vim.o.shiftwidth = tonumber(input)
+        end)
+<
```
Comment (gpanders) [inline runtime/doc/lua.txt:2555]: Example snippet got separated from `Example: ` text and placed after parameters??

File: runtime/doc/lua.txt [hunk 15]
```diff
@@ -2587,6 +2600,22 @@ vim.ui.select({items}, {opts}, {on_choice})                  *vim.ui.select()*
...
+                return "I'd like to choose " .. item
+            end,
+        }, function(choice)
+            if choice == 'spaces' then
+                vim.o.expandtab = true
+            else
+                vim.o.expandtab = false
+            end
+        end)
+<
```
Comment (gpanders) [inline runtime/doc/lua.txt:2617]: same thing here

File: runtime/doc/lua.txt [hunk 16]
```diff
@@ -3049,35 +3076,37 @@ A version "range spec" defines a semantic version range which can be
...
-   Partial right: missing pieces treated as x (2.3 => 2.3.x).
-   1.2.3 - 2.3       is >=1.2.3 <2.4.0
-   1.2.3 - 2         is >=1.2.3 <3.0.0
-
-   Partial left: missing pieces treated as 0 (1.2 => 1.2.0).
-   1.2 - 2.3.0       is 1.2.0 - 2.3.0
+versions (1.2.3-rc1) are not matched.
+
+>
+
```
Comment (gpanders) [inline runtime/doc/lua.txt:3082]: newline

File: runtime/doc/treesitter.txt [hunk 17]
```diff
@@ -921,8 +921,11 @@ list_predicates()                     *vim.treesitter.query.list_predicates()*
 omnifunc({findstart}, {base})                *vim.treesitter.query.omnifunc()*
     Omnifunc for completing node names and predicates in treesitter queries.
 
-    Use via >lua
-      vim.bo.omnifunc = 'v:lua.vim.treesitter.query.omnifunc'
+    Use via
+
+    >lua
+
+        vim.bo.omnifunc = 'v:lua.vim.treesitter.query.omnifunc'
```
Comment (gpanders) [inline runtime/doc/treesitter.txt:928]: newline

File: runtime/doc/lua.txt (hunk 8 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: 2d8f5c1802ea4c0a36c12fd4709dad4a0cc25c54

File: runtime/doc/lsp.txt (hunk 4 above)
Comment (gpanders) [reply runtime/doc/lsp.txt]: bd1f77687055c4d74736e6923b4da0b6739110ed

File: runtime/doc/lsp.txt (hunk 3 above)
Comment (gpanders) [reply runtime/doc/lsp.txt]: bd1f77687055c4d74736e6923b4da0b6739110ed

File: runtime/doc/lua.txt (hunk 9 above)
Comment (gpanders) [reply runtime/doc/lua.txt:797]: 02734df2f9ad329ee2bb371f8ce632bfd7cd2b65

Comment (gpanders) [reply runtime/doc/lua.txt]: 02734df2f9ad329ee2bb371f8ce632bfd7cd2b65

File: runtime/doc/lsp.txt (hunk 5 above)
Comment (gpanders) [reply runtime/doc/lsp.txt]: a611e660ace8ee669a623e5650936c0a4ca804a2

File: runtime/doc/lua.txt (hunk 7 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: a611e660ace8ee669a623e5650936c0a4ca804a2

File: runtime/doc/lua.txt (hunk 11 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: a611e660ace8ee669a623e5650936c0a4ca804a2

File: runtime/doc/lua.txt (hunk 16 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: a611e660ace8ee669a623e5650936c0a4ca804a2

File: runtime/doc/treesitter.txt (hunk 17 above)
Comment (gpanders) [reply runtime/doc/treesitter.txt]: a611e660ace8ee669a623e5650936c0a4ca804a2

File: runtime/doc/lsp.txt (hunk 6 above)
Comment (gpanders) [reply runtime/doc/lsp.txt]: 93878bf87c2e61ecaaa51cf7cb2f39cc227137b7

File: runtime/doc/lua.txt (hunk 10 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: 93878bf87c2e61ecaaa51cf7cb2f39cc227137b7

File: runtime/doc/lua.txt (hunk 12 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: a611e660ace8ee669a623e5650936c0a4ca804a2

File: runtime/doc/api.txt (hunk 1 above)
Comment (gpanders) [reply runtime/doc/api.txt]: 95c7ca533c502cc66f249f71ea657f26d616ec1f

File: runtime/doc/lua.txt (hunk 14 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: 89838a0fa4ef2f6d7e8adb97e09d136f0fb9f940

File: runtime/doc/lua.txt (hunk 15 above)
Comment (gpanders) [reply runtime/doc/lua.txt]: 89838a0fa4ef2f6d7e8adb97e09d136f0fb9f940

File: runtime/doc/api.txt [hunk 18]
```diff
@@ -3055,13 +3067,24 @@ nvim_open_win({buffer}, {enter}, {*config})                  *nvim_open_win()*
...
 
     Example (Lua): window-relative float >lua
-        vim.api.nvim_open_win(0, false,
-          {relative='win', row=3, col=3, width=12, height=3})
+
+        vim.api.nvim_open_win(0, false, {
+          relative='win',
+          row=3,
+          col=3,
+          width=12,
```
Comment (justinmk) [inline runtime/doc/api.txt:3075]: vertical space in docs is precious, people will often need to compare things and anticipate following text without jumping around. i don't think the old format was hard to read. and the code itself is usually not the main purpose of the examples, rather it's to show the shape of something.

**Comment by justinmk** [comment]
>  In cases where we have a code sample inside of a list, a `` ``` `` code block doesn't work (at least, I can't figure out how to make it work). Example:
> 
> https://github.com/neovim/neovim/blob/27a566f3f8e07a4cebb426674800bdf9a7f4f222/runtime/lua/vim/diagnostic.lua#L600-L610

this is what I understand the markdown to be for that:

```
 ---                       * format: (function) A function that takes a diagnostic as input and 
 ---                                 returns a string. The return value is the text used to display 
 ---                                 the diagnostic. Example: 
 ---                         ```
 ---                         function(diagnostic) 
 ---                           if diagnostic.severity == vim.diagnostic.severity.ERROR then 
 ---                             return string.format("E: %s", diagnostic.message) 
 ---                           end 
 ---                           return diagnostic.message 
 ---                         end 
 ---                         ```
```

if that doesn't work then yeah we're stuck :(

File: runtime/doc/api.txt (hunk 18 above)
Comment (gpanders) [reply runtime/doc/api.txt]: Reverted 91b17a5ee9a5f77845779d2198cf15625a556bb1

File: runtime/doc/api.txt [hunk 19]
```diff
@@ -2553,8 +2563,9 @@ nvim_buf_get_extmarks({buffer}, {ns_id}, {start}, {end}, {*opts})
     Region can be given as (row,col) tuples, or valid extmark ids (whose
     positions define the bounds). 0 and -1 are understood as (0,0) and (-1,-1)
     respectively, thus the following are equivalent: >lua
-      vim.api.nvim_buf_get_extmarks(0, my_ns, 0, -1, {})
-      vim.api.nvim_buf_get_extmarks(0, my_ns, {0,0}, {-1,-1}, {})
+
+        vim.api.nvim_buf_get_extmarks(0, my_ns, 0, -1, {})
```
[the remark this answers is not in the document]
Comment (gpanders) [reply runtime/doc/api.txt:2567]: We can, though personally I prefer it (it serves as a useful visual break between text and code sample).

**Comment by gpanders** [comment]
> this is what I understand the markdown to be for that:
> 
> ```
>  ---                       * format: (function) A function that takes a diagnostic as input and 
>  ---                                 returns a string. The return value is the text used to display 
>  ---                                 the diagnostic. Example: 
>  ---                         ```
>  ---                         function(diagnostic) 
>  ---                           if diagnostic.severity == vim.diagnostic.severity.ERROR then 
>  ---                             return string.format("E: %s", diagnostic.message) 
>  ---                           end 
>  ---                           return diagnostic.message 
>  ---                         end 
>  ---                         ```
> ```
> 
> if that doesn't work then yeah we're stuck :(

Indeed, and that is what I tried. This is the resulting diff to the .txt file:

```diff
diff --git a/runtime/doc/diagnostic.txt b/runtime/doc/diagnostic.txt
index c37ecb265..bdd3166aa 100644
--- a/runtime/doc/diagnostic.txt
+++ b/runtime/doc/diagnostic.txt
@@ -416,15 +416,11 @@ config({opts}, {namespace})                          *vim.diagnostic.config()*
                          to render an LSP diagnostic error code.
                        • format: (function) A function that takes a diagnostic
                          as input and returns a string. The return value is
-                         the text used to display the diagnostic. Example: >lua
-
-                             function(diagnostic)
-                               if diagnostic.severity == vim.diagnostic.severity.ERROR then
-                                 return string.format("E: %s", diagnostic.message)
-                               end
-                               return diagnostic.message
-                             end
-<
+                         the text used to display the diagnostic. Example:
+                         `lua function(diagnostic) if diagnostic.severity ==
+                         vim.diagnostic.severity.ERROR then return
+                         string.format("E: %s", diagnostic.message) end return
+                         diagnostic.message end `
 
                      • signs: (default true) Use signs for diagnostics.
                        Options:
```

I tried every permutation I could think of. None worked.

File: scripts/gen_eval_files.lua [hunk 20]
```diff
@@ -213,17 +213,20 @@ end
 
 --- Convert vimdoc references to markdown literals
 --- Convert vimdoc codeblocks to markdown codeblocks
+---
+--- Ensure code blocks have one empty line before the start fence and after the closing fence.
```
Comment (justinmk) [inline scripts/gen_eval_files.lua:217]: why? not in favor of this

File: runtime/doc/api.txt (hunk 19 above)
[the remark this answers is not in the document]
Comment (justinmk) [reply runtime/doc/api.txt:2567]: Sometimes it's good but sometimes it's way too much, so I think we should let the doc writer make the decision.

[the remark this answers is not in the document]
Comment (gpanders) [reply runtime/doc/api.txt:2560]: Removed 0d1e04e12a6a1f899404981a411a004992d2bd9d

File: scripts/gen_eval_files.lua (hunk 20 above)
Comment (justinmk) [inline scripts/gen_eval_files.lua:217]: is this comment still relevant?

**Review by justinmk — APPROVED** [review]
Lots of nice improvements in the rendered docs

File: scripts/gen_eval_files.lua (hunk 20 above)
Comment (gpanders) [reply scripts/gen_eval_files.lua:217]: Yes. Note this is only for the Markdown code blocks that are autogenerated in the `_meta/` files (e.g. `_meta/api.lua`). It ensures that Markdown code blocks have the form:

````
  --- Lorem ipsum etc. etc.
  ---
  --- Example:
> ---                 < -- blank line here!
  --- ```
  --- ...
  --- ...
  --- ```
> ---                 < -- blank line here!
  ---@param foo string Foo
  ---@return table
````

This is not strictly required by the spec, but in my own experience I've found that omitting these blank lines causes issues with some renderers in some cases (including GitHub and `gen_vimdoc.py`).

**Pull request merged**